internal/v2/fakes2av2/fakes2av2.go

```diff
@@ -53,6 +64,79 @@ func (s *Server) SetUpSession(stream s2av2pb.S2AService_SetUpSessionServer) erro
 	}
 }
 
+// TODO(rmehta19): consider also returning an error with this function.
+func validatePeerCertificateChain(req *s2av2pb.ValidatePeerCertificateChainReq) (*s2av2pb.SessionResp, error) {
+	switch x := req.PeerOneof.(type) {
+	case *s2av2pb.ValidatePeerCertificateChainReq_ClientPeer_:
+		// TODO(rmehta19): Handle Verification Mode(UNSPECIFIED, SPIFFE, or CONNECT_TO_GOOGLE).
+		derCertChain := req.GetClientPeer().CertificateChain
+		if len(derCertChain) == 0 {
+			// TODO(rmehta19): Is it an error for peer to have empty cert chain(I think
+			// so, because we set ClientAuth as require request and verify)? If so,
+			// (server requires nonempty certificate chain) will error be raised
+			// for us by some existing API function way before this peer verification?
+			err := errors.New("Client Peer Verification failed: client cert chain is empty.")
+			return buildValidatePeerCertificateChainSessionResp(400, err.Error(), s2av2pb.ValidatePeerCertificateChainResp_FAILURE, err.Error()), err
+		}
+
+		// Obtain the set of root certificates.
+		rootCertPool := x509.NewCertPool()
+		if ok := rootCertPool.AppendCertsFromPEM(clientCert); ok != true {
+			err := errors.New("Client Peer Verification failed: S2Av2 could not obtain/parse roots.")
+			return buildValidatePeerCertificateChainSessionResp(500, err.Error(), s2av2pb.ValidatePeerCertificateChainResp_FAILURE, err.Error()), err
+		}
+
+		// Verify the leaf certificate.
+		opts := x509.VerifyOptions{
+			CurrentTime: time.Now(),
+			Roots: rootCertPool,
+		}
+		x509LeafCert, err := x509.ParseCertificate(derCertChain[0])
+		if err != nil {
+			s := fmt.Sprintf("Client Peer Verification failed: %v", err)
+			return buildValidatePeerCertificateChainSessionResp(400, s, s2av2pb.ValidatePeerCertificateChainResp_FAILURE, s), err
+		}
+		if _, err := x509LeafCert.Verify(opts); err != nil {
+			s := fmt.Sprintf("Client Peer Verification failed: %v", err)
+			return buildValidatePeerCertificateChainSessionResp(400, s, s2av2pb.ValidatePeerCertificateChainResp_FAILURE, s), err
+		}
+		return buildValidatePeerCertificateChainSessionResp(200, "", s2av2pb.ValidatePeerCertificateChainResp_SUCCESS, "Client Peer Verification succeeded"), nil
+	case *s2av2pb.ValidatePeerCertificateChainReq_ServerPeer_:
+		// TODO(rmehta19): Handle Verification Mode(UNSPECIFIED, SPIFFE, or CONNECT_TO_GOOGLE).
+		derCertChain := req.GetServerPeer().CertificateChain
```

Comment:
Can we move the verification logic into a helper function? The code is a bit complex so I think doing so would be helpful. :)